rama-tls-acme/src/client.rs

```diff
@@ -0,0 +1,664 @@
+    pub async fn nonce(&self) -> Result<String, OpaqueError> {
+        if let Some(nonce) = self.nonce.lock().take() {
+            return Ok(nonce);
+        }
+
+        let response = self
+            .https_client
+            .head(&self.directory.new_nonce)
+            .send(Context::default())
+            .await
+            .context("fetch new nonce")?;
+
+        let nonce = Self::get_nonce_from_response(&response)?;
+        Ok(nonce)
+    }
+
+    fn get_nonce_from_response(response: &Response) -> Result<String, OpaqueError> {
+        Ok(response
+            .header_str(REPLAY_NONCE_HEADER)
+            .context("get nonce from headers")?
+            .to_owned())
+    }
+
+    /// Create a new account with the given [`CreateAccountOptions`] options
+    pub async fn create_account(
+        &self,
+        options: CreateAccountOptions,
+    ) -> Result<Account, ClientError> {
+        let key = EcdsaKey::generate().context("generate key for account")?;
+
+        let do_request = async || {
+            let response = self
+                .post(&self.directory.new_account, Some(&options), &key)
+                .await
+                .context("create account request")?;
+
+            let location: String = response
+                .header_str(Location::name())
+                .context("get location header")?
+                .into();
+
+            let account = parse_response::<server::Account>(response).await?;
+            Ok((location, account))
+        };
+
+        let (location, account) = retry_bad_nonce(do_request).await?;
+
+        Ok(Account {
+            client: self,
+            inner: account,
+            credentials: AccountCredentials { key, kid: location },
+        })
+    }
+
+    async fn post(
+        &self,
+        url: &str,
+        payload: Option<&impl Serialize>,
+        signer: &impl Signer,
+    ) -> Result<Response, OpaqueError> {
+        let mut builder = JWSBuilder::new();
+        if let Some(payload) = payload {
+            let payload = serde_json::to_vec(payload).context("serialize payload")?;
+            builder.set_payload(payload);
+        }
+
+        let nonce = self.nonce().await?;
+        builder.try_set_protected_headers(ProtectedHeaderAcme {
+            nonce: Cow::Owned(nonce),
+            url: Cow::Borrowed(url),
+        })?;
+
+        let jws = builder.build_flattened(signer)?;
+
+        let request = self
+            .https_client
+            .post(url)
+            .typed_header(UserAgent::from_static("rama-tls-acme"))
+            .typed_header(ContentType::jose_json())
+            .json(&jws);
+
+        let response = request.send(Context::default()).await?;
+
+        *self.nonce.lock() = Some(Self::get_nonce_from_response(&response)?);
+        Ok(response)
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Enum of popular acme providers and their directory url
+pub enum AcmeProvider {
+    LetsEncrypt(&'static str),
```

Comment:
Why is this an enum if you anyway have their directory url as a public field? I would either just make it a simple enum or have it not be an enum at all but instead an empty type or so with const's.

Reply:
I would just have it as 

```
pub enum AcmeProvider {
    LetsEncrypt,
    SomethingElse,
    Foo,
}
```

If you have something bedies the nromal production one you can still add another entry for that one (e.g. LetsEncryptTest )or w/e... But bit weird to have that str there, especially as it is public, so I can just add there whatever I want even if it's not LetsEncrypt :P